### PR TITLE
skip hpugraph usage for first token to save memory

### PR DESCRIPTION
**Pull request by polisettyvarma**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by puneeshkhanna — APPROVED** [review]

**Comment by polisettyvarma** [comment]
please review this @dvarshney-habana

File: examples/text-generation/README.md [hunk 1]
```diff
@@ -77,6 +77,7 @@ Here are a few settings you may be interested in:
 - `--batch_size` to specify the batch size
 - `--bf16` to run generation in bfloat16 precision (or to be specified in your DeepSpeed configuration if using DeepSpeed)
 - `--use_hpu_graphs` to use [HPU graphs](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_HPU_Graphs.html) to speed up generation
+- `--limit_hpu_graphs` to skip HPUGraph usage for first token to save memory
```
Comment (regisss) [inline examples/text-generation/README.md:80]:
```suggestion
- `--limit_hpu_graphs` to skip HPU Graph usage for first token to save memory
```

File: examples/text-generation/run_generation.py [hunk 2]
```diff
@@ -166,6 +166,11 @@ def main():
         type=str,
         help="Output directory to store results in.",
     )
+    parser.add_argument(
+        "--limit_hpu_graphs",
+        action="store_true",
+        help="Skip HPUGraph usage for first token to save memory",
```
Comment (regisss) [inline examples/text-generation/run_generation.py:172]:
```suggestion
        help="Skip HPU Graph usage for first token to save memory",
```

File: optimum/habana/transformers/generation/utils.py [hunk 3]
```diff
@@ -432,6 +434,11 @@ def generate(
         if streamer is not None:
             streamer.put(input_ids.cpu())
 
+        # determine whether limit_hpu_graphs needs to be used
+        model_kwargs["limit_hpu_graphs"] = (
+            generation_config.limit_hpu_graphs if hasattr(generation_config, "limit_hpu_graphs") else False
+        )
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:440]: Let's define `limit_hpu_graphs` in: https://github.com/huggingface/optimum-habana/blob/4bf8a224b8838850e67ac85ca7fdbca91cc09d2d/optimum/habana/transformers/generation/configuration_utils.py#L19
so that we don't have to check here if it's in the generation config.

File: optimum/habana/transformers/generation/utils.py [hunk 4]
```diff
@@ -1096,12 +1103,20 @@ def greedy_search(
             # prepare model inputs
             model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
 
+            hpu_graphs_kwargs = {}
+            if model_kwargs["limit_hpu_graphs"]:
+                hpu_graphs_kwargs.update({"bypass_hpu_graphs": False})
+                if "first_token" not in model_kwargs.keys():
+                    model_kwargs["first_token"] = True
+                    hpu_graphs_kwargs.update({"bypass_hpu_graphs": True})
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1111]: Maybe we could define a `get_hpu_graphs_kwargs` method in this class to factorize this block which is called in every decoding method?
So that we have:
```py
hpu_graphs_kwargs = get_hpu_graphs_kwargs(model_kwargs)
```
It will be easier to maintain.

File: optimum/habana/transformers/generation/utils.py [hunk 5]
```diff
@@ -1408,12 +1423,20 @@ def sample(
...
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1431]:
```suggestion
            hpu_graphs_kwargs = get_hpu_graphs_kwargs(model_kwargs)
```

File: optimum/habana/transformers/generation/utils.py [hunk 6]
```diff
@@ -1726,11 +1749,19 @@ def beam_search(
 
             model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
 
+            hpu_graphs_kwargs = {}
+            if model_kwargs["limit_hpu_graphs"]:
+                hpu_graphs_kwargs.update({"bypass_hpu_graphs": False})
+                if "first_token" not in model_kwargs.keys():
+                    model_kwargs["first_token"] = True
+                    hpu_graphs_kwargs.update({"bypass_hpu_graphs": True})
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1757]:
```suggestion
            hpu_graphs_kwargs = get_hpu_graphs_kwargs(model_kwargs)
```

File: optimum/habana/transformers/generation/utils.py [hunk 7]
```diff
@@ -2370,11 +2401,19 @@ def constrained_beam_search(
...
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:2409]:
```suggestion
            hpu_graphs_kwargs = get_hpu_graphs_kwargs(model_kwargs)
```

File: optimum/habana/transformers/generation/utils.py (hunk 4 above)
Comment (polisettyvarma) [reply optimum/habana/transformers/generation/utils.py:1111]: image

i did actually but when i run style it changes the code which is wrong and i need to add on top of it something
i can unpack tuple then dict

File: optimum/habana/transformers/generation/utils.py (hunk 5 above)
Comment (polisettyvarma) [reply optimum/habana/transformers/generation/utils.py:1431]: same as above

File: optimum/habana/transformers/generation/utils.py (hunk 6 above)
Comment (polisettyvarma) [reply optimum/habana/transformers/generation/utils.py:1757]: same as above

File: optimum/habana/transformers/generation/utils.py (hunk 7 above)
Comment (polisettyvarma) [reply optimum/habana/transformers/generation/utils.py:2409]: same as above

File: optimum/habana/transformers/generation/utils.py (hunk 4 above)
Comment (regisss) [reply optimum/habana/transformers/generation/utils.py:1111]: Oh that's weird! What do you return in `_get_hpu_graphs_kwargs`?

**Comment by polisettyvarma** [comment]
please review now @regisss, @dvarshney-habana and @puneeshkhanna

File: optimum/habana/transformers/generation/configuration_utils.py
Comment (puneeshkhanna) [inline optimum/habana/transformers/generation/configuration_utils.py:1]: add doc string too similar to attn softmax bf16

Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/configuration_utils.py:1]: Rest all looks good to me

**Review by dvarshney-habana — APPROVED** [review]

**Review by regisss — COMMENTED** [review]
CI failed because `_get_hpu_graphs_kwargs` should be linked to Transformers' `generate here: https://github.com/huggingface/optimum-habana/blob/af3da936c338907c446b4612e649eafee147c5a3/optimum/habana/transformers/modeling_utils.py#L103

Or, maybe easier, you can just declare this method outside of this class and use it that way without `self`.

**Comment by polisettyvarma** [comment]
> CI failed because `_get_hpu_graphs_kwargs` should be linked to Transformers' `generate here:
> 
> https://github.com/huggingface/optimum-habana/blob/af3da936c338907c446b4612e649eafee147c5a3/optimum/habana/transformers/modeling_utils.py#L103
> 
> Or, maybe easier, you can just declare this method outside of this class and use it that way without `self`.

I took the first approach, please review @regisss

File: optimum/habana/transformers/generation/utils.py [hunk 8]
```diff
@@ -131,6 +131,14 @@ def _expand_dict_for_generation(dict_to_expand):
 
         return input_ids, model_kwargs
 
+    def _get_hpu_graphs_kwargs(self, model_kwargs):
+        hpu_graphs_kwargs = {}
+        if model_kwargs["limit_hpu_graphs"]:
+            hpu_graphs_kwargs.update({"bypass_hpu_graphs": False})
+            if "first_token" not in model_kwargs.keys():
+                model_kwargs["first_token"] = True
+                hpu_graphs_kwargs.update({"bypass_hpu_graphs": True})
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:140]: We also need to return `hpu_graphs_kwargs` at the end

**Review by regisss — COMMENTED** [review]
@polisettyvarma CI failed, we need to return `hpu_graphs_kwargs` at the end of `_get_hpu_graphs_kwargs`.
There is also one merge conflict to solve, probably due to the trim logit PR that I just merged. There shouldn't be much to do I think.

**Comment by polisettyvarma** [comment]
please review now @regisss

**Review by regisss — APPROVED** [review]

**Comment by yafshar** [comment]
@regisss, @polisettyvarma which model is using this feature? I do not see `bypass_hpu_graphs` in any models' arguments. Would you please point me to a model which is using this?

**Comment by regisss** [comment]
> @regisss, @polisettyvarma which model is using this feature? I do not see `bypass_hpu_graphs` in any models' arguments. Would you please point me to a model which is using this?

@yafshar You won't see it in the code of any model. It's managed by the HPU graph API which wraps the forward of the model.
For example, in a Docker container running the SynapseAI 1.15 Docker image, you can find it at the beginning of the `wrapped_hpugraph_forward` method in `/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/graphs.py`.
